File: C++/0752. Open the Lock.cpp
class Solution {
public:
    int openLock(vector<string>& deadends, string target) {
        queue<string> q;
        set<string> s(deadends.begin(), deadends.end());

        if(target == "0000") {
            return 0;
        }

        if(find(deadends.begin(), deadends.end(), "0000") != deadends.end()) {
            return -1;
        }

        q.push("0000");
        int num = 0;
        while(!q.empty()) {
            num++;
            int l = q.size();
            for(int k = 0; k < l; k++) {
                string temp = q.front();
                q.pop();
                for(auto delta:{1, 9}) {
                    for(int i = 0; i < 4; ++i) {
                        string tt = temp.substr(0, i) + char((temp[i] - '0' + delta) % 10 + '0') + temp.substr(i + 1, 4 - i -1);
                        if(tt == target) {
                            return num;
                        }
                        if(s.count(tt) == 0) {
                            q.push(tt);
                            s.insert(tt);
                        }
                    }
                }
            }
        }
        return -1;
    }
};

File: Python3/0752. Open the Lock.py
class Solution:
    def openLock(self, deadends: List[str], target: str) -> int:
        q = [('0000',0)]

        deadends = set(deadends)

        if '0000' in deadends:
            return -1

        if target == '0000':
            return 0

        while q:
            num, times = q.pop(0)
            for delta in [1, 9]:
                for i in range(4):
                    new_num = num[:i] + str((int(num[i]) + delta) % 10) + num[i + 1:]
                    if new_num == target:
                        return times + 1
                    if new_num not in deadends:
                        q.append((new_num, times + 1))
                        deadends.add(new_num)
                    
        return -1
